### PR TITLE
Etcm 22 structured definitions for net gas metering

### PR DESCRIPTION
**Pull request by mmrozek**
https://eips.ethereum.org/EIPS/eip-2200

### CODE REVIEW

File: src/test/scala/io/iohk/ethereum/vm/SSTOREOpCodeGasPostConstantinopleSpec.scala [hunk 1]
```diff
@@ -10,47 +10,76 @@ import akka.util.ByteString.{empty => bEmpty}
...
+      ("60006000556002600055", 1, 5812, 0),
+      ("60026000556000600055", 1, 5812, 15000),
+      ("60026000556003600055", 1, 5812, 0),
+      ("60026000556001600055", 1, 5812, 4200),
+      ("60026000556002600055", 1, 5812, 0),
+      ("60016000556000600055", 1, 5812, 15000),
+      ("60016000556002600055", 1, 5812, 0),
+      ("60016000556001600055", 1, 1612, 0),
+      ("600160005560006000556001600055", 0, 40818, 19200),
+      ("600060005560016000556000600055", 1, 10818, 19200)
```
Comment (KonradStaniec) [inline src/test/scala/io/iohk/ethereum/vm/SSTOREOpCodeGasPostConstantinopleSpec.scala:68]: I found out that geth has two additional test cases : https://github.com/ethereum/go-ethereum/blob/b9df7ecdc3d3685180ceb29665bab59e9f614da5/core/vm/gas_table_test.go#L76. 
```
	{1, 2306, "0x6001600055", 2306, 0, ErrOutOfGas},                            // 1 -> 1 (2300 sentry + 2xPUSH)
	{1, 2307, "0x6001600055", 806, 0, nil},                                     // 1 -> 1 (2301 sentry + 2xPUSH)
```
 Most probably testing some edge cases. Imo we should try to add them, as all this gass couting makes me nervous,

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala [hunk 2]
```diff
@@ -593,52 +595,67 @@ case object MSTORE8 extends OpCode(0x53, 2, 0, _.G_verylow) {
...
-              state.config.feeSchedule.G_sreset - state.config.feeSchedule.G_sload
+    val currentBlockNumber = state.env.blockHeader.number
+    val etcFork = state.config.blockchainConfig.etcForkForBlockNumber(currentBlockNumber)
+    val ethFork = state.config.blockchainConfig.ethForkForBlockNumber(currentBlockNumber)
+
+    val eip2200Enabled = isEip2200Enabled(etcFork, ethFork)
+    val eip1283Enabled = isEip1283Enabled(ethFork)
+
+    if(eip2200Enabled && state.gas <= state.config.feeSchedule.G_callstipend){
+        state.withError(OutOfGas)
```
Comment (KonradStaniec) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala:606]: Usually `OutOfGas` errors zeros caller remaining gas like ` state.copy(gas = 0).withError(OutOfGas)`. Do you think we should try to do the same here ? (maybe it worth checking parity of geth)

File: src/test/scala/io/iohk/ethereum/vm/SSTOREOpCodeGasPostConstantinopleSpec.scala (hunk 1 above)
Comment (mmrozek) [reply src/test/scala/io/iohk/ethereum/vm/SSTOREOpCodeGasPostConstantinopleSpec.scala]: Added

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala (hunk 2 above)
Comment (mmrozek) [reply src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: The tests from previous comment check that

**Review by KonradStaniec — APPROVED** [review]

**Pull request merged**